File: scripts/governance/check_mutation_score.py
#!/usr/bin/env python3
"""Mutation Testing Gate — 70% mutation score on changed Python files"""

import subprocess
import sys

THRESHOLD = 70


def get_changed_files() -> list[str]:
    try:
        result = subprocess.run(
            ["git", "diff", "--cached", "--name-only", "--diff-filter=ACMR"],
            capture_output=True,
            text=True,
        )
        return [
            f
            for f in result.stdout.strip().split("\n")
            if f.startswith("app/") and f.endswith(".py")
        ]
    except Exception:
        return []


def has_mutmut() -> bool:
    try:
        subprocess.run(["mutmut", "--version"], capture_output=True)
        return True
    except FileNotFoundError:
        return False


changed = get_changed_files()
if not changed:
    print("✓ Mutation score: no source files changed")
    sys.exit(0)

if not has_mutmut():
    print("⚠ Mutation score: mutmut not installed — skip (pip install mutmut)")
    sys.exit(0)

# Run mutmut on changed files
paths_arg = " ".join(f"--paths-to-mutate={f}" for f in changed)
result = subprocess.run(
    f"mutmut run {paths_arg} --no-progress", shell=True, capture_output=True, text=True
)

# Parse result
results_output = subprocess.run(
    ["mutmut", "results"], capture_output=True, text=True
).stdout

# Simple score estimate from output
killed = results_output.count("🎉")
survived = results_output.count("🙁")
total = killed + survived

if total == 0:
    print("✓ Mutation score: no mutants generated")
    sys.exit(0)

score = (killed / total) * 100
print(f"  Mutants: {total}, Killed: {killed}, Score: {score:.1f}%")

if score < THRESHOLD:
    print(f"✗ Mutation score: {score:.1f}% < {THRESHOLD}% threshold", file=sys.stderr)
    sys.exit(1)

print(f"✓ Mutation score: {score:.1f}% meets {THRESHOLD}% threshold")

File: scripts/governance/check-mutation-score.js
#!/usr/bin/env node
/**
 * Mutation Testing Gate — 70% mutation score on changed files.
 * Uses StrykerJS when available. Falls back to reporting only.
 */

const fs = require('fs');
const { execSync } = require('child_process');

const THRESHOLD = 70;

function getChangedFiles() {
  try {
    return execSync('git diff --cached --name-only --diff-filter=ACMR', { encoding: 'utf8' })
      .trim().split('\n')
      .filter(f => f.startsWith('src/') && /\.(ts|tsx)$/.test(f) && !f.endsWith('.d.ts'));
  } catch { return []; }
}

const changedFiles = getChangedFiles();

if (changedFiles.length === 0) {
  console.log('✓ Mutation score: no source files changed');
  process.exit(0);
}

// Check if stryker is available
const hasStryker = (() => {
  try {
    execSync('npx stryker --version', { stdio: 'pipe' });
    return true;
  } catch { return false; }
})();

if (!hasStryker) {
  console.log(`⚠ Mutation score: StrykerJS not installed — skip (install: npm install --save-dev @stryker-mutator/core @stryker-mutator/typescript-checker)`);
  process.exit(0);
}

// Run stryker on changed files
const mutatePattern = changedFiles.join(',');
try {
  const output = execSync(
    `npx stryker run --mutate "${mutatePattern}" --reporters json --logLevel error`,
    { encoding: 'utf8', timeout: 120000 }
  );

  const reportFile = 'reports/mutation/mutation.json';
  if (fs.existsSync(reportFile)) {
    const report = JSON.parse(fs.readFileSync(reportFile, 'utf8'));
    const score = report.mutationScore ?? 0;
    if (score < THRESHOLD) {
      console.error(`✗ Mutation score: ${score.toFixed(1)}% < ${THRESHOLD}% threshold`);
      process.exit(1);
    }
    console.log(`✓ Mutation score: ${score.toFixed(1)}% meets ${THRESHOLD}% threshold`);
  }
} catch (e) {
  console.log(`⚠ Mutation score: StrykerJS run failed — ${e.message.split('\n')[0]}`);
  process.exit(0);
}

process.exit(0);
